cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+
+## 6 -- Packaging Rules
+#### Libraries
+1. It is discouraged to have small packages (with couple of cc and header files).
+2. Functionality used by multiple packages should go in the `.../src` and `.../interface` directories.
+3. Only header files that expose a public interface should go into the `.../interface` directory.  
+4. Include only files that are in the current directory, e.g.  
+`#include "some_header.h"`  
+ or in the `.../interface directory`, e.g  
+`#include "Subsystem/Package/interface/some_header.h"`
+#### Plugins
+5. Put plugins (e.g. EDProducers, EDAnalyzers, etc.) into a `Package/Subpackage/plugins/` directory, with its dedicated `BuildFile.xml`.
+6. Do not split plugins into header and source files. If you do split them, keep the header files in the `.../plugins` directory.
+7. All code used only by the plugins in `../plugins` should also go under `.../plugins`.
+ `plugins.cc` and `SealPlugins.cc` or any special files that just define plugins, except for template instantiations, are discouraged. 
+#### Tests
+8. Unit tests to test the functionalities of your Library and/or Plugins should go under `.../test`.
+9. Add test library/plugins in `.../test/BuildFile.xml` for the common functionality used only by your unit tests.
+10. For unit tests which simply run `cmsRun your-cfg` (to test your plugin), please use `<test name="..." command="cmsRun …"/>` in your `.../test/BuildFile.xml`.
+#### Python
+11. A `_cfi` file should contain only the definition of one module, and possible Modifier ("era") customizations on it. The module label should be the same as the `_cfi` file name.
+12. The `_cfi` file should be left to be generated automatically with the `fillDescriptions()`. When Modifier customizations are needed, the auto-generated label should have e.g. "Default" postfix and be imported+cloned to the desired name.
+13. A module/Task/Sequence/Path with a given name should be defined in exactly one `_cfi` or `_cff` file.
+14. All Modifier customizations on a module/Task/Sequence/Path should be applied on the same file that defines the module/Task/Sequence/Path.
+15. When one customizes an existing parameter in `clone()`, `Modifier.toModify()`, or in assignment, explicit types on the right hand side should be avoided.
+#### Data files
+16. To keep the repository size under control, we discourage adding any data files under `.../data`.
+17. Please use cms-data externals github repositories to add/change any data file.
+#### Binaries and scripts
+18. Public executables/binaries should go under `.../bin`.
+19. It is discouraged to generate plugins from `.../bin`.
+20. Additional libraries used only by multiple executables of your `.../bin` should also go under `.../bin`.
+21. Any scripts/utilities which should be available publicly (i.e. in PATH) should go under `.../bin` and use the `INSTALL_SCRIPTS` flag in `.../bin/BuildFile.xml`.
+22. Adding scripts under `.../scripts` is discouraged.
```

Comment:
Maybe worth elaborating on why the method described in rule 21 should be preferred vs. using `.../scripts`?

Reply:
I don't know the reason. Maybe the author of these lines could comment.

Reply:
We have seen that user basically dump many things in scripts directory (scripts, root macros,  configuration files, input data to be used etc.). This pollutes the PATH as every thing from scripts sub-dir is blindly copied in to top level bin. We can change rule to have something like
```
22. Only add scripts which should be available in PATH in `.../scripts` directory. Keep configuration and data files in separate directories e.g data
```

Reply:
This does seem more sensible than having scripts under `bin` if they don't need to be compiled.

Reply:
Yes, I am in favor of the proposed change.

Reply:
I re-worded these rules along the lines @smuzaffar suggested.